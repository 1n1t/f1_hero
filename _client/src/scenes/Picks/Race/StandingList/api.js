/* eslint-disable import/prefer-default-export */
import fetchApi from '../../../../services/api';

const endPoints = {
	postPick(round) {
		return `/api/picks/${round}`;
	}
};

export const postPick = (round, payload) =>
	fetchApi(endPoints.postPick(round), 'post', payload);
