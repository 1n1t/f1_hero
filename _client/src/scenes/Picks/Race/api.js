/* eslint-disable import/prefer-default-export */
import fetchApi from '../../../services/api';

const endPoints = {
	getPick(round) {
		return `/api/picks/${round}`;
	}
};

export const getPick = round => fetchApi(endPoints.getPick(round));
