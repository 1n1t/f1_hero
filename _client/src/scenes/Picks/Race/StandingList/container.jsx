/* eslint-disable react/prop-types */
import React from 'react';
import { connect } from 'react-redux';

import StandingList from './index';
import { updateRace } from '../../../../data/races/actions';

const StandingListContainer = props => <StandingList {...props} />;

export const mapStateToProps = state => ({
	races: state.data.races.all,
	drivers: state.data.drivers.all
});

export default connect(mapStateToProps, { updateRace })(StandingListContainer);
