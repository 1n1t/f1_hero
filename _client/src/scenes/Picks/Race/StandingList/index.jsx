/* eslint-disable jsx-a11y/click-events-have-key-events */
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { arrayMove } from 'react-sortable-hoc';
import { calculatePickAndStemLeftPosition } from 'App/services/utils/functions';

import ForecastNew from './ForecastNew';
import ForecastMissed from './ForecastMissed';
import ForecastSubmitted from './ForecastSubmitted';
import SubmitButton from './SubmitButton';
import { postPick } from './api';
import { racePropType } from '../index';
import './styles.scss';

class StandingList extends Component {
	state = {
		pickPos: 0,
		stemPos: 0,
		list:
			this.props.status === 'submitted' ? this.props.list : this.props.drivers,
		status: this.props.status
	};

	componentDidMount() {
		const { races, round, raceNode, raceHolderNode } = this.props;
		const { pickPos, stemPos } = calculatePickAndStemLeftPosition(
			{
				picksNode: this.picks,
				stemNode: this.stem,
				raceNode,
				raceHolderNode
			},
			races.length,
			round
		);

		const newState = {
			...this.state,
			pickPos,
			stemPos
		};
		this.setState(newState);
	}

	onSortEnd = ({ oldIndex, newIndex }) => {
		const newList = arrayMove(this.state.list, oldIndex, newIndex);
		const formattedList = newList.map((value, index) => ({
			...value,
			position: index + 1
		}));

		const newState = {
			...this.state,
			list: formattedList //	todo make list array immutable
		};
		this.setState(newState);
	};

	submitPick = async () => {
		const { round, onSubmit, updateRace } = this.props;
		try {
			const { pick } = (await postPick(round, {
				round,
				forecast: this.state.list.slice(0, 10)
			})).data;

			const newState = {
				...this.state,
				status: 'submitted',
				list: pick.forecast
			};

			this.setState(newState);
			onSubmit(pick);
			updateRace({ round, field: 'hasPick', value: true });
		} catch (error) {
			//	todo display an error as a flash message
			console.error('error: unable to save pick!!', error);
		}
	};

	render() {
		const { status } = this.state;
		let Standings;

		switch (status) {
			case 'new':
				Standings = ForecastNew;
				break;
			case 'missed':
				Standings = ForecastMissed;
				break;
			default:
				Standings = ForecastSubmitted;
		}

		return (
			<div
				ref={el => {
					this.picks = el;
				}}
				className="StandingList"
				style={{ left: `${this.state.pickPos}px` }}
			>
				<div
					ref={el => {
						this.stem = el;
					}}
					className="stem"
					style={{ left: `${this.state.stemPos}px` }}
				/>
				<Standings
					items={this.state.list}
					onSortEnd={this.onSortEnd}
					lockToContainerEdges
					helperClass="draggable"
					lockAxis="y"
				/>
				{status === 'new' && <SubmitButton onClick={this.submitPick} />}
			</div>
			//	todo make function that will create refs
		);
	}
}

export const listItemPropTypes = {
	position: PropTypes.number.isRequired,
	_drivers: PropTypes.arrayOf(
		PropTypes.shape({
			code: PropTypes.string.isRequired,
			name: PropTypes.string.isRequired
		})
	)
};

StandingList.defaultProps = {
	list: [],
	drivers: [],
	onSubmit: () => {}
};

StandingList.propTypes = {
	status: PropTypes.oneOf(['new', 'submitted', 'passed']).isRequired,
	list: PropTypes.arrayOf(PropTypes.shape(listItemPropTypes)),
	drivers: PropTypes.arrayOf(PropTypes.shape(listItemPropTypes)),
	round: PropTypes.number.isRequired,
	races: PropTypes.arrayOf(PropTypes.shape(racePropType)).isRequired,
	onSubmit: PropTypes.func,
	updateRace: PropTypes.func.isRequired,
	// eslint-disable-next-line react/forbid-prop-types
	raceHolderNode: PropTypes.object.isRequired,
	// eslint-disable-next-line react/forbid-prop-types
	raceNode: PropTypes.object.isRequired
};

export default StandingList;
