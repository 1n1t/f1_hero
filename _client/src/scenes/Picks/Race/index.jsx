/* eslint-disable jsx-a11y/click-events-have-key-events */
import React, { Component } from 'react';
import PropTypes from 'prop-types';

import { isObjectEmpty, makeRef } from 'App/services/utils/functions';
import { getPick } from './api';
import StandingList from './StandingList/container';
import StatusIcon from './StatusIcon';
import './styles.scss';

class Race extends Component {
	state = {
		pick: {}
	};

	setRacePick = pick => {
		const newState = {
			...this.state,
			pick: {
				...this.state.pick,
				...pick
			}
		};

		this.setState(newState);
	};

	fetchPick = async () => {
		const { round } = this.props.race;
		this.props.selectRace(round);

		const { pick } = (await getPick(round)).data;
		this.setRacePick(pick);
	};

	displayPick() {
		const { race, selectedRace, raceHolderNode } = this.props;
		const { pick } = this.state;
		const isPickSet = !isObjectEmpty(pick);
		const isRaceSelected = selectedRace.round === race.round;

		if (!isRaceSelected || (!isPickSet && race.hasPick)) {
			return null;
		}

		if (race.hasPick) {
			return (
				<StandingList
					list={pick.forecast}
					status="submitted"
					round={race.round}
					raceNode={this.raceNode}
					raceHolderNode={raceHolderNode}
				/>
			);
		} else if (race.isPassed) {
			return (
				<StandingList
					status="missed"
					round={race.round}
					raceNode={this.raceNode}
					raceHolderNode={raceHolderNode}
				/>
			);
		}
		return (
			<StandingList
				status="new"
				round={race.round}
				onSubmit={this.setRacePick}
				raceNode={this.raceNode}
				raceHolderNode={raceHolderNode}
			/>
		);
	}

	displayStatusIcon() {
		const { isPassed, hasPick } = this.props.race;

		// eslint-disable-next-line no-nested-ternary
		return isPassed ? (
			<StatusIcon iconClass="fa-check" status="passed" />
		) : hasPick ? (
			<StatusIcon iconClass="fa-circle" status="submitted" />
		) : (
			<StatusIcon iconClass="fa-circle-o" status="new" />
		);
	}

	render() {
		const { flagUrl, alpha3code, score } = this.props.race;
		const scoreText = score ? `${score}pt` : '-';

		return (
			<div
				ref={makeRef('raceNode', this)}
			>
				<div
					className="Race"
					onClick={this.fetchPick}
					role="link"
					tabIndex="0"
					data-test="race"
				>
					{this.displayStatusIcon()}
					<div className="img-container">
						<img src={flagUrl} alt={`flag of ${alpha3code}`} />
					</div>
					<p className="country-code">{alpha3code}</p>
					<p className="score">{scoreText}</p>
				</div>
				{this.displayPick()}
			</div>
		);
	}
}

export const racePropType = {
	round: PropTypes.number.isRequired,
	isPassed: PropTypes.bool.isRequired,
	hasPick: PropTypes.bool.isRequired,
	flagUrl: PropTypes.string.isRequired,
	alpha3code: PropTypes.string.isRequired,
	score: PropTypes.number
};

Race.propTypes = {
	race: PropTypes.shape(racePropType).isRequired,
	selectedRace: PropTypes.shape({
		round: PropTypes.number.isRequired
	}).isRequired,
	selectRace: PropTypes.func.isRequired,
	// eslint-disable-next-line react/forbid-prop-types
	raceHolderNode: PropTypes.object.isRequired
};

export default Race;
