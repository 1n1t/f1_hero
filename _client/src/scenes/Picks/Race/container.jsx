/* eslint-disable react/prop-types */
import React from 'react';
import { connect } from 'react-redux';

import Race from './index';
import { selectRace } from '../../../data/races/actions';

const RaceContainer = props => <Race {...props} />;

export const mapStateToProps = state => ({
	selectedRace: state.data.races.selected
});

export default connect(mapStateToProps, { selectRace })(RaceContainer);
