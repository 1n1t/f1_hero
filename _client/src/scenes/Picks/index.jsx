import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { makeRef } from 'App/services/utils/functions';

import Race from './Race/container';
import Loader from '../../components/Loader';
import './styles.scss';

class Picks extends Component {
	state: {
		isMounted: false
	};

	componentDidMount() {
		this.setState({ isMounted: true });
	}

	renderContent() {
		const { races } = this.props;

		return races && races.length > 0 ? (
			races.map(race => (
				<Race
					key={race.round}
					race={race}
					races={races}
					raceHolderNode={this.raceHolderNode}
				/>
			))
		) : (
			<Loader />
		);
	}

	render() {
		return (
			<div className="Picks" ref={makeRef('raceHolderNode', this)}>
				{this.renderContent()}
			</div>
		);
	}
}

Picks.propTypes = {
	races: PropTypes.arrayOf(
		PropTypes.shape({
			round: PropTypes.number
		})
	).isRequired
};

export default Picks;
