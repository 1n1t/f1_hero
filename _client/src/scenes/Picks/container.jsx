/* eslint-disable react/prop-types */
import React, { Component } from 'react';
import { connect } from 'react-redux';
import { updatePageHeading } from '../../services/page/actions';
import { getRaceList } from '../../data/races/actions';
import { getDriverList } from '../../data/drivers/actions';
import Picks from './index';

class PicksContainer extends Component {
	componentDidMount() {
		this.props.updatePageHeading('My picks');
		this.props.getRaceList();
		this.props.getDriverList();
	}

	render() {
		return <Picks races={this.props.races} />;
	}
}

const mapStateToProps = state => ({ races: state.data.races.all });

export default connect(mapStateToProps, {
	updatePageHeading,
	getRaceList,
	getDriverList
})(PicksContainer);
