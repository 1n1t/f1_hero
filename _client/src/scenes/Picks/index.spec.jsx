import r from 'Root/jest/utils/renderer';
import racesSample from 'JestClient/samples/races';
import pickSample from 'JestClient/samples/pick';
import { sel } from 'JestClient/functions';
import moxios from 'moxios';

import Loader from 'App/components/Loader';
import Picks from './index';
import Race from './Race/container';
import StandingList from './Race/StandingList/container';

const defaultProps = {
	races: racesSample.all
};

r.init(Picks, defaultProps);

it('renders correctly', () => {
	const wrapper = r.render(null, true);
	expect(wrapper).toMatchSnapshot();
});

it('renders correct amount of races', () => {
	const wrapper = r.render(null, true);
	const races = wrapper.find(Race);

	expect(races).toHaveLength(defaultProps.races.length);
});

it('renders Loader if there are no races', () => {
	const wrapper = r.render({ races: [] }, true);
	const loader = wrapper.find(Loader);
	const races = wrapper.find(Race);

	expect(loader).toHaveLength(1);
	expect(races).toHaveLength(0);
});

describe('renders standing list', () => {
	it('only if race is clicked', done => {
		const wrapper = r.render(null, false);
		let standingList = wrapper.find(StandingList);
		expect(standingList).toHaveLength(0);

		const index = 0;
		const race = wrapper.find(sel('race')).at(index);

		moxios.withMock(() => {
			race.simulate('click');

			moxios.wait(async () => {
				const request = moxios.requests.mostRecent();

				await request.respondWith({
					status: 200,
					response: { pick: pickSample }
				});

				standingList = wrapper.update().find(StandingList);
				expect(standingList).toHaveLength(1);
				done();
			});
		});
	});
});
