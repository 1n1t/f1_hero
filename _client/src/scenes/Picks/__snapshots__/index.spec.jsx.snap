// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="Picks"
>
  <Connect(RaceContainer)
    key="1"
    race={
      Object {
        "alpha3code": "AUS",
        "flagUrl": "http://foo.bar",
        "hasPick": true,
        "isPassed": true,
        "round": 1,
        "score": 50,
      }
    }
    races={
      Array [
        Object {
          "alpha3code": "AUS",
          "flagUrl": "http://foo.bar",
          "hasPick": true,
          "isPassed": true,
          "round": 1,
          "score": 50,
        },
        Object {
          "alpha3code": "RUS",
          "flagUrl": "http://foo.bar",
          "hasPick": false,
          "isPassed": true,
          "round": 2,
          "score": 0,
        },
        Object {
          "alpha3code": "SPA",
          "flagUrl": "http://foo.bar",
          "hasPick": true,
          "isPassed": false,
          "round": 3,
          "score": 0,
        },
        Object {
          "alpha3code": "BEL",
          "flagUrl": "http://foo.bar",
          "hasPick": false,
          "isPassed": false,
          "round": 4,
          "score": 0,
        },
      ]
    }
  />
  <Connect(RaceContainer)
    key="2"
    race={
      Object {
        "alpha3code": "RUS",
        "flagUrl": "http://foo.bar",
        "hasPick": false,
        "isPassed": true,
        "round": 2,
        "score": 0,
      }
    }
    races={
      Array [
        Object {
          "alpha3code": "AUS",
          "flagUrl": "http://foo.bar",
          "hasPick": true,
          "isPassed": true,
          "round": 1,
          "score": 50,
        },
        Object {
          "alpha3code": "RUS",
          "flagUrl": "http://foo.bar",
          "hasPick": false,
          "isPassed": true,
          "round": 2,
          "score": 0,
        },
        Object {
          "alpha3code": "SPA",
          "flagUrl": "http://foo.bar",
          "hasPick": true,
          "isPassed": false,
          "round": 3,
          "score": 0,
        },
        Object {
          "alpha3code": "BEL",
          "flagUrl": "http://foo.bar",
          "hasPick": false,
          "isPassed": false,
          "round": 4,
          "score": 0,
        },
      ]
    }
  />
  <Connect(RaceContainer)
    key="3"
    race={
      Object {
        "alpha3code": "SPA",
        "flagUrl": "http://foo.bar",
        "hasPick": true,
        "isPassed": false,
        "round": 3,
        "score": 0,
      }
    }
    races={
      Array [
        Object {
          "alpha3code": "AUS",
          "flagUrl": "http://foo.bar",
          "hasPick": true,
          "isPassed": true,
          "round": 1,
          "score": 50,
        },
        Object {
          "alpha3code": "RUS",
          "flagUrl": "http://foo.bar",
          "hasPick": false,
          "isPassed": true,
          "round": 2,
          "score": 0,
        },
        Object {
          "alpha3code": "SPA",
          "flagUrl": "http://foo.bar",
          "hasPick": true,
          "isPassed": false,
          "round": 3,
          "score": 0,
        },
        Object {
          "alpha3code": "BEL",
          "flagUrl": "http://foo.bar",
          "hasPick": false,
          "isPassed": false,
          "round": 4,
          "score": 0,
        },
      ]
    }
  />
  <Connect(RaceContainer)
    key="4"
    race={
      Object {
        "alpha3code": "BEL",
        "flagUrl": "http://foo.bar",
        "hasPick": false,
        "isPassed": false,
        "round": 4,
        "score": 0,
      }
    }
    races={
      Array [
        Object {
          "alpha3code": "AUS",
          "flagUrl": "http://foo.bar",
          "hasPick": true,
          "isPassed": true,
          "round": 1,
          "score": 50,
        },
        Object {
          "alpha3code": "RUS",
          "flagUrl": "http://foo.bar",
          "hasPick": false,
          "isPassed": true,
          "round": 2,
          "score": 0,
        },
        Object {
          "alpha3code": "SPA",
          "flagUrl": "http://foo.bar",
          "hasPick": true,
          "isPassed": false,
          "round": 3,
          "score": 0,
        },
        Object {
          "alpha3code": "BEL",
          "flagUrl": "http://foo.bar",
          "hasPick": false,
          "isPassed": false,
          "round": 4,
          "score": 0,
        },
      ]
    }
  />
</div>
`;
